### DISCUSSION TITLE
Why my code is not working?

### DISCUSSION THREAD

**Discussion by charlesneimog** [DBraun/DawDreamer#55] [Q&A] [answered]
Hi,

I am starting with python, I am not able to understand how the graph work.
```

import dawdreamer as daw
import numpy as np
from scipy.io import wavfile
import librosa

SAMPLE_RATE = 44100
BUFFER_SIZE = 512
DURATION = 15

def load_audio_file(file_path, duration=None):
  sig, rate = librosa.load(file_path, duration=duration, mono=False, sr=SAMPLE_RATE)
  assert(rate == SAMPLE_RATE)
  return sig

engine = daw.RenderEngine(SAMPLE_RATE, BUFFER_SIZE)
meu_plugin = engine.make_plugin_processor("plugin", "C:/Users/neimog/OneDrive_usp.br/Documents/Plugins/VST3/dearVR MICRO.vst3")

meu_plugin.set_parameter(0, 45) # override a specific parameter.
print(meu_plugin.get_parameter_name(0))

multi =  load_audio_file("C:/USERS/NEIMOG/ONEDRIVE_USP.BR/DOCUMENTS/OpenMusic/OUT-FILES/om-ckn/Fl-mul-G#5-G4+-D6-100-cents.wav-v-stereo.wav")

graph = [
  (meu_plugin, [multi])] ##  

engine.load_graph(graph)
engine.render(DURATION)
audio = engine.get_audio()  

wavfile.write('my_song2.wav', SAMPLE_RATE, audio.transpose()) # don't forget to transpose!

```
How make the audio `multi `be processed by `meu_plugin`

Thank you very much!!!!!

**Comment by DBraun** [comment, marked as answer]
The main issue is that you need to create a playback processor with `engine.make_playback_processor("playback", multi)`
I also think that `meu_plugin.set_parameter(0, 45)` won't work well because VST parameters must be between 0 and 1. 45 is outside this range.

Full script:
```python
import dawdreamer as daw
import numpy as np
from scipy.io import wavfile
import librosa

SAMPLE_RATE = 44100
BUFFER_SIZE = 512
DURATION = 15

def load_audio_file(file_path, duration=None):
  sig, rate = librosa.load(file_path, duration=duration, mono=False, sr=SAMPLE_RATE)
  assert(rate == SAMPLE_RATE)
  return sig

engine = daw.RenderEngine(SAMPLE_RATE, BUFFER_SIZE)
meu_plugin = engine.make_plugin_processor("plugin", "C:/Users/neimog/OneDrive_usp.br/Documents/Plugins/VST3/dearVR MICRO.vst3")

meu_plugin.set_parameter(0, 45) # override a specific parameter.
print(meu_plugin.get_parameter(0))  # warning, is this actually 45? Probably not. VST parameters are normally between 0 and 1.
print(meu_plugin.get_parameter_name(0))

multi =  load_audio_file("C:/USERS/NEIMOG/ONEDRIVE_USP.BR/DOCUMENTS/OpenMusic/OUT-FILES/om-ckn/Fl-mul-G#5-G4+-D6-100-cents.wav-v-stereo.wav")

playback_processor = engine.make_playback_processor("playback", multi)

graph = [
  (playback_processor , []),  # playback takes no inputs
  (meu_plugin, ["playback"])
]

## this also works
#graph = [
#  (playback_processor , []),  # playback takes no inputs
#  (meu_plugin, [playback_processor.get_name()])
#]

engine.load_graph(graph)
engine.render(DURATION)
audio = engine.get_audio()  

wavfile.write('my_song2.wav', SAMPLE_RATE, audio.transpose()) # don't forget to transpose!
```

**Reply by Felix-Eger** [reply]
Hi, 

I am currently working on a project where I am only interested to render short audio snippets produced by a synthesizer. 
My synthesizer of choice at this moment is Sylenth, which comes as a dll and provides .fxp-files as presets. 

Unfortunately I am quite stuck when trying to render any audio from my synthesizer and any example that I found isn't really working for me. The strange aspect is, I can render an actual audio file, however, there simply isn't any sound. 

Would be great to point out what I need to do to get it working!

(I am including an audio file in the processing though that's not even needed for my task...)

`SYNTH_PLUGIN = "C:/Program Files/Steinberg/VstPlugins/Sylenth1.dll"
SAMPLE_RATE = 44100
BUFFER_SIZE = 64

path = "some_audio.wav"

def load_audio_file(file_path, duration=None):
    sig, rate = librosa.load(file_path, duration=duration, mono=False, sr=SAMPLE_RATE)
    assert(rate == SAMPLE_RATE)
    return sig

engine = daw.RenderEngine(SAMPLE_RATE, BUFFER_SIZE)
synth = engine.make_plugin_processor("my_synth", SYNTH_PLUGIN)
multi = load_audio_file(path)
playback_preprocessor = engine.make_playback_processor("abc", multi)

graph = [
    (playback_preprocessor,[]),
    (synth,[playback_preprocessor.get_name()])
]

engine.load_graph(graph)
engine.render(DURATION)
audio = engine.get_audio()

wavfile.write('test_123.wav', SAMPLE_RATE, audio.transpose())
`

**Reply by DBraun** [reply]
Hopefully the only issue is that you need to specify MIDI notes:
```python
synth.add_midi_note(67, 127, 0.5, .25) # (MIDI note, velocity, start sec, duration sec)
# or
synth.load_midi("some_file.mid")
```

As you noted, the playback_preprocessor is unnecessary because most synthesizers don't take audio input.

**Comment by charlesneimog** [comment]
great,

Thank you!
